### PR TITLE
Convert REST Schema/ DTOs to Records, improve Swagger descs

### LINKED ISSUE #2264: Change SGv2/REST DTOs to be simple Record types where possible
(from #2213 code review)

Many (if not all) of DTOs used by REST API could be replaced by Java 15 Record types: this would simplify code .

### PR DESCRIPTION
**Pull request by tatu-at-datastax**
**What this PR does**:

Now that Stargate V2 uses JDK 17 we can use Record types for DTOs to reduce boilerplate.
May also make minor improvements to swagger/OpenApi annotations.

**Which issue(s) this PR fixes**:
Fixes #2264

**Checklist**
- [x] Changes manually tested
- [ ] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java [hunk 1]
```diff
@@ -71,129 +45,61 @@ public TableOptions getTableOptions() {
...
     }
 
     public boolean hasPartitionKey(String key) {
-      return (partitionKey != null) && partitionKey.contains(key);
+      return partitionKey.contains(key);
     }
 
     public boolean hasClusteringKey(String key) {
-      return (clusteringKey != null) && clusteringKey.contains(key);
+      return clusteringKey.contains(key);
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java:69]: null checks removed? they are still existing on likes 56-57

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java [hunk 2]
```diff
@@ -71,129 +45,61 @@ public TableOptions getTableOptions() {
...
-    }
   }
 
   // copied from SGv1 ClusteringExpression
-  public static class ClusteringExpression {
+  public record ClusteringExpression(
+      @Schema(required = true, description = "The name of the column to order by") String column,
+      // 18-Jul-2022, tatu: Originally with DropWizard, Swagger 3, we had:
+      //    allowableValues = "ASC,DESC"
+      // but hoping "enumeration" works as replacement
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java:94]: does it work? And what's the difference from `allowableValues`?

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java [hunk 3]
```diff
@@ -5,82 +5,47 @@
...
-  public void setName(String name) {
+public record Sgv2TableAddRequest(
+    @Schema(
+            required = true,
+            description = "The name of the table to add.",
+            example = "cycling_events")
+        @NotBlank(message = "TableAdd.name must be provided")
+        String name,
+    @Schema(
+            required = true,
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java:16]: if this is required you should add `@NotNull`

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java [hunk 4]
```diff
@@ -5,82 +5,47 @@
...
+            description =
+                "The primary key definition of the table, consisting of partition and clustering keys.")
+        Sgv2Table.PrimaryKey primaryKey,
+    @Schema(
+            required = true,
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java:21]: same here, `@NotNull`, maybe even `@NotEmpty` make sense..

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTAddRequest.java [hunk 5]
```diff
@@ -1,48 +1,16 @@
...
-    return fields;
-  }
-
-  public void setFields(List<Sgv2UDT.UDTField> fields) {
-    this.fields = fields;
-  }
-}
+public record Sgv2UDTAddRequest(
+    @Schema(required = true, description = "User Defined Type name") String name,
+    @Schema(required = true, description = "User Defined Type fields")
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTAddRequest.java:11]: same validation missing if required?

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTUpdateRequest.java [hunk 6]
```diff
@@ -1,88 +1,22 @@
...
-  }
-
-  public void setRenameFields(List<FieldRename> renameFields) {
-    this.renameFields = renameFields;
-  }
+public record Sgv2UDTUpdateRequest(
+    @Schema(required = true, description = "User Defined Type name")
+        @NotBlank(message = "UserDefinedTypeUpdate.name must be provided")
+        String name,
+    @Schema(description = "User Defined Type fields to add") List<Sgv2UDT.UDTField> addFields,
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTUpdateRequest.java:14]: Can this be empty?

File: apis/sgv2-restapi/src/test/java/io/stargate/sgv2/it/RestApiV2QIntegrationTestBase.java [hunk 7]
```diff
@@ -288,33 +287,30 @@ protected void createKeyspace(String keyspaceName) {
...
             new Sgv2ColumnDefinition("lastName", "text", false),
             new Sgv2ColumnDefinition("firstName", "text", false),
-            new Sgv2ColumnDefinition("age", "int", false)));
-
-    Sgv2Table.PrimaryKey primaryKey = new Sgv2Table.PrimaryKey();
-    primaryKey.setPartitionKey(Arrays.asList("id"));
-    tableAdd.setPrimaryKey(primaryKey);
+            new Sgv2ColumnDefinition("age", "int", false));
 
+    Sgv2Table.PrimaryKey primaryKey = new Sgv2Table.PrimaryKey(Arrays.asList("id"), null);
```
Comment (ivansenic) [inline apis/sgv2-restapi/src/test/java/io/stargate/sgv2/it/RestApiV2QIntegrationTestBase.java:297]: nit: maybe add constructor that only accepts the list?

**Review by ivansenic — COMMENTED** [review]
Left some comments.. Not sure if tests do have actual changes, but moving those to rest-based assertions would be great, because we would not need to check the "entities" now, but simply confirm that the JSON is correct.. Maybe somebody else should have another look, don't feel confident to approve myself..

**Comment by tatu-at-datastax** [comment]
> Left some comments.. Not sure if tests do have actual changes, but moving those to rest-based assertions would be great, because we would not need to check the "entities" now, but simply confirm that the JSON is correct.. Maybe somebody else should have another look, don't feel confident to approve myself..

Yeah I was trying to balance amount of change to do to imrpove things vs. limited time before DocsV3 work.
Changes should be mechanical and although it'd be nice to have unit tests (not just integration) I felt I need to wrap this up.

I'll address other comments and I think that I would value your approval -- but take the responsibility if something changed i.e. not expect you to find issues. Just provide sanity checking.

**Comment by ivansenic** [comment]
@tatu-at-datastax Forgot to push the update? There was no changes since I first reviewed it..

**Comment by tatu-at-datastax** [comment]
> @tatu-at-datastax Forgot to push the update? There was no changes since I first reviewed it..

No, I just had to focus on other work. So no fixes yet, adding comments before realizing I need to switch.

Will cc you when I go back to making changes here, after getting something first done for DocsV3/shredding.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2ColumnDefinition.java [hunk 8]
```diff
@@ -4,91 +4,38 @@
...
+            required = true,
+            description = "Name for the column, which must be unique.")
+        @NotBlank(message = "columnDefinition.name must be provided")
+        String name,
+    @Schema(
+            example = "text",
+            required = true,
+            description = "The type of data allowed in the column.")
+        String typeDefinition,
+    @Schema(description = "Denotes whether the column is shared by all rows of a partition.")
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2ColumnDefinition.java:26]: > should we add `defaultValue=false` here?

Yes, makes sense, will add.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RESTResponse.java [hunk 9]
```diff
@@ -15,31 +15,11 @@
...
-  }
-
-  @JsonCreator
-  public Sgv2RESTResponse(@JsonProperty("data") final T data) {
-    this.data = data;
-  }
-}
+@Schema(name = "RESTResponseWrapper")
+public record Sgv2RESTResponse<T>(
+    @Schema(description = "Response data returned by the request.") T data) {}
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RESTResponse.java:25]: I don't think `null` makes sense so will add.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RowsResponse.java [hunk 10]
```diff
@@ -15,26 +15,19 @@
...
-      @JsonProperty("pageState") String pageState,
-      @JsonProperty("rows") List<Map<String, Object>> rows) {
-    super(count, pageState, rows);
-  }
-}
+@Schema(name = "RowsResponse")
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public record Sgv2RowsResponse(
+    @Schema(description = "The count of records returned.") int count,
+    @Schema(
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RowsResponse.java:29]: Yes, thank you!

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RowsResponse.java [hunk 11]
```diff
@@ -15,26 +15,19 @@
...
+            description =
+                "A string representing the paging state to be used on future paging requests.")
+        String pageState,
+    @Schema(description = "The rows returned by the request.") List<Map<String, Object>> data) {}
```
[the remark this answers is not in the document]
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2RowsResponse.java:33]: Unfortunately I think that might be the best we can do: it is a JSON Array of JSON Objects, with varying column names, types.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java (hunk 2 above)
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2Table.java:94]: I think it does based on comments I saw back in the day: will remove the comment -- new property is the replacement for the old one.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTUpdateRequest.java (hunk 6 above)
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2UDTUpdateRequest.java:14]: Yes, both can be empty.

File: apis/sgv2-restapi/src/test/java/io/stargate/sgv2/it/RestApiV2QIntegrationTestBase.java (hunk 7 above)
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/test/java/io/stargate/sgv2/it/RestApiV2QIntegrationTestBase.java:297]: Added.

**Comment by tatu-at-datastax** [comment]
@ivansenic Went ahead and made suggested changes -- thank you for reviewing -- now going through CI.

File: apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java (hunk 3 above)
Comment (tatu-at-datastax) [reply apis/sgv2-restapi/src/main/java/io/stargate/sgv2/restapi/service/models/Sgv2TableAddRequest.java:19]: Ok after trying it out realized that Update Table does not require primary key or columns; only create does.

**Review by maheshrajamani — APPROVED** [review]

**Pull request merged**